src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -528,26 +528,31 @@ std::unique_ptr<Uptane::Target> SotaUptaneClient::findTargetInDelegationTree(con
   return findTargetHelper(*toplevel_targets, target, 0, false);
 }
 
-result::Download SotaUptaneClient::downloadImages(const std::vector<Uptane::Target> &targets,
+result::Download SotaUptaneClient::downloadImages(std::vector<Uptane::Target> targets,
                                                   const api::FlowControlToken *token) {
   // Uptane step 4 - download all the images and verify them against the metadata (for OSTree - pull without
   // deploying)
   std::lock_guard<std::mutex> guard(download_mutex);
   result::Download result;
   std::vector<Uptane::Target> downloaded_targets;
-  for (auto it = targets.cbegin(); it != targets.cend(); ++it) {
-    auto images_target = findTargetInDelegationTree(*it);
+  for (auto &it : targets) {
+    auto images_target = findTargetInDelegationTree(it);
     if (images_target == nullptr) {
       // TODO: Could also be a missing target or delegation expiration.
-      last_exception = Uptane::TargetMismatch(it->filename());
-      LOG_ERROR << "No matching target in images targets metadata for " << *it;
+      last_exception = Uptane::TargetMismatch(it.filename());
+      LOG_ERROR << "No matching target in images targets metadata for " << it;
       result = result::Download(downloaded_targets, result::DownloadStatus::kError, "Target hash mismatch.");
       sendEvent<event::AllDownloadsComplete>(result);
       return result;
     }
+    // If the URL from the Director is unset, but the URL from the Images repo
+    // is set, use that.
+    if (it.uri().empty() && !images_target->uri().empty()) {
+      it.setUri(images_target->uri());
+    }
   }
-  for (auto it = targets.cbegin(); it != targets.cend(); ++it) {
-    auto res = downloadImage(*it, token);
+  for (const auto &it : targets) {
+    auto res = downloadImage(it, token);
```

Comment:
for some reason downloadImage() passes a target object by value, maybe it makes sense to pass it by reference ?

Reply:
This was changed because the targets inside the vectors are modified with `setUri()` and we probably don't want to change it in the vector that lives in the caller.

That's a bit annoying though, I would maybe suggest passing by const reference and making a copy inside the function so that it's more explicit.

Reply:
>  I would maybe suggest passing by const reference and making a copy inside the function so that it's more explicit. (+ a comment about that)

Agreed, done.